clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java

```diff
@@ -1030,6 +1030,7 @@ public void assign(Collection<TopicPartition> partitions) {
      * @return The fetched records (may be empty)
      */
     private Map<TopicPartition, List<ConsumerRecord<K, V>>> pollOnce(long timeout) {
+        client.maybeTriggerWakeup();
```

Comment:
@ijuma but in that case, wouldn't the poll not trigger that exception potentially due to `pollNoWakeup` not being secured by some locking.

In 
```java
    public void pollNoWakeup() {
        disableWakeups();
        try {
            poll(0, time.milliseconds(), null);
        } finally {
            enableWakeups();
        }
    }
```

I have no guarantee that I don't `poll` right after `disableWakeups` making the poll not throw  by happening before `enableWakeups`?

I feel like even with the above solution I'd still have to somehow lock throughout `pollNoWakeup` and around 

```java
client.poll(0, ...)
```

in which case my above solution of just locking on the wakeup atomic boolean via:

```java
    public void maybeTriggerWakeup() {
        synchronized (this.wakeup) {
            if (wakeupDisabledCount == 0 && wakeup.get()) {
                log.trace("Raising wakeup exception in response to user wakeup");
                wakeup.set(false);
                throw new WakeupException();
            }
        }
    }
```

may be a little easier to follow than some locking outside `ConsumerNetworkClient` ?

Summary :) : I don't think simply checking if data is available is enough, we do have to make sure we don't have the wakeup disabled when calling `poll`.